File: Solutions/Parenthesis Game/code.cpp
#include<bits/stdc++.h> 
using namespace std; 


int Array( int arr1[] , int arr2[] , char* str , int n ) 
{ 

	stack<int> stk; 
	int result = 0; 

	for (int i=0; i<n; i++) 					//logic for a valid bracket pair !!!!!
	{ 
		
		if (str[i] == '(') 
		stk.push(i); 

		else 
		{
			if (!stk.empty()) 					//important!!!!!!
			{ 
				arr2[i] = 1; 
				arr1[stk.top()] = 1; 
				stk.pop(); 
			}
			else
				arr2[i] = 0; 
		} 
	} 
	

	for( int i = 1 ; i < n; i++ ) 
	{ 
		result +=arr2[i];
	} 
	
	return result*2;
} 


int main() 
{ 

	char str[100000]; 
	cin>>str;
	int n = strlen(str); 

	int arr1[n+1] ={0}; 		
	int arr2[n+1] ={0}; 		
	
        cout<<Array( arr1 , arr2 , str, n )<<endl; 

	return 0; 
} 

File: Solutions/Parenthesis Game/code.java
import java.io.*;
import java.util.*;
 
class balgame
{
 static int Array( int arr1[] , int arr2[] , char str[] , int n,Stack<Integer>stk)
{
int result = 0; 
 for (int i=0; i<n; i++) 					
	{ 
	if (str[i] == '(') 
		stk.push(i); 
 
		else 
		{
			if (!stk.empty()) 					//important af
			{ 
				arr2[i] = 1; 
				arr1[(Integer)stk.peek()] = 1; 
				stk.pop(); 
			} 
 
	 
			else
				arr2[i] = 0; 
		} 
	} 
	
 
	for( int i = 1 ; i < n; i++ ) 
	{ 
		result +=arr2[i];
	} 
	
	return result*2;
} 
 
 
public static void main(String args[]) 
{ 
 
	String s;
   Scanner sc=new Scanner(System.in);
   Stack<Integer> stk=new Stack<Integer>();
   s=sc.next();
   int n,i;
   n=s.length();
   char []str=s.toCharArray();
 int arr1[]=new int[n+1];
 int arr2[]=new int[n+1];
 for(i=0;i<n+1;i++)
	arr1[i] =0;
	for(i=0;i<n+1;i++) 
	arr2[i] =0; 	
	
    System.out.println(Array( arr1 , arr2 , str, n,stk )); 
 
 
}
}
